src/LendController.sol

```diff
@@ -0,0 +1,156 @@
+    }
+
+    uint256 ONE = 1e18;
+    mapping(address => mapping(address => bool)) public termsApprovals;
+    mapping(address => mapping(address => uint256)) public lenderBalance;
+    mapping(uint256 => mapping(address => uint256)) public lenderLoanShares;
+
+    INFTLoanFacilitator public loanFacilitator;
+
+    constructor(INFTLoanFacilitator _facilitator) {
+        loanFacilitator = _facilitator;
+    }
+
+    event SetTermsApproval(address indexed from, address indexed termsContract, bool approved);
+
+    function setTermsApproval(address termsContract, bool approved) external {
+        termsApprovals[msg.sender][termsContract] = approved;
+
+        emit SetTermsApproval(msg.sender, termsContract, approved);
+    }
+
+    function purchaseNFT() external {
+        // allows anyone to call in and seize a seize NFT that
+        // the controller holds the lend ticket for and pays the lenders
+        // purchase price must be >= totalOwed on loan
+    }
+
+    function liquidateNFT(uint256 loanId) external {
+        // TBD if controller will have one liquidation mechanism 
+        // or will call out to the terms contract for implement
+        // update lenderBalance based on lenderLoanShares * sale value 
+    }
+
+    // used to lend to an exisiting loan
+    function lend(
+        uint256 loanId,
+        uint256 tokenId,
+        address nft,
+        BorrowRequest memory request,
+        bytes calldata data
+    ) external {
+        _lend(loanId, tokenId, nft, request, false, data);
+    }
+
+    function _lend(
+        uint256 loanId,
+        uint256 tokenId,
+        address nft,
+        BorrowRequest memory request,
+        bool skipIsBuyoutCheck,
+        bytes calldata data
+    ) internal {
+        LoanTerms memory terms = NFTLoanTermsSource(request.termsSource).terms(tokenId, msg.sender, request.terms.loanAsset, '');
+
+        ERC20 loanAsset = ERC20(request.terms.loanAsset);
+        uint256 lenderTotal;
+
+        for(uint i = 0; i < request.lendersSpecs.length; i++) {
+            LenderSpec memory info = request.lendersSpecs[i];
+            require(termsApprovals[info.lender][request.termsSource], 'lender has not approved terms source');
+
+            if(info.pull) {
+                ERC20(request.terms.loanAsset).safeTransferFrom(info.lender, address(this), info.amount);
+            } else {
+                lenderBalance[info.lender][address(loanAsset)] -= info.amount;
+            }
+            lenderTotal += info.amount;
+
+            lenderLoanShares[loanId][info.lender] = info.amount * ONE / terms.amount;
+        }
+
+        loanAsset.approve(address(loanFacilitator), type(uint256).max);
+
+        if(skipIsBuyoutCheck){
+            require(lenderTotal == terms.amount);
+        } else {
+            (,,, uint40 lastAccumulatedTimestamp,,,,,,,) = loanFacilitator.loanInfo(loanId);
+            
+            if(lastAccumulatedTimestamp != 0) {
+                require(terms.allowBuyouts);
```

Comment:
need to think if this is really the best way to do this / what we want to do. Alternatively could 
- `require(terms.amount >= loanFacilitator.totalOwed(loanId))`
- or just not do any check? If the terms contract is willing to lend to this asset, it should be willing to buyout? Expecting to get buyout amount back on repayment. (Off chain will have to ensure that `terms` include a 10% improvement over existing).

Reply:
i think cool to not do any check. borrow ticket holder could then use UI to see if theres any way they could get better terms on their loan

Reply:
well but this issue is, if you're a lender, are you OK more than X of your funds moving because you need to payback the previous lender, in addition to lending?